### PR TITLE
New aarch64 insn support

### PR DESCRIPTION
**Pull request by jj16791**
Various new aarch64 instruction support from recent ModSim work.

### CODE REVIEW

**Review by FinnWilkinson — COMMENTED** [review]
Generally all looks good. Just a few minor nit-picks

**Review by rahahahat — APPROVED** [review]

File: src/include/simeng/arch/aarch64/helpers/neon.hh [hunk 1]
```diff
@@ -889,6 +931,28 @@ class neonHelp {
 
     return {out, 256};
   }
+
+  /** Helper function for NEON instructions with the format `zip<1,2> vd.T,
+   * vn.T, vm.T`.
+   * T represents the type of operands (e.g. for vn.d, T = uint64_t).
+   * I represents the number of operands (e.g. for vn.8b, I = 8).
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/arch/aarch64/helpers/neon.hh]: Updated

File: src/include/simeng/arch/aarch64/helpers/neon.hh [hunk 2]
```diff
@@ -690,6 +707,31 @@ class neonHelp {
     return {out, 256};
   }
 
+  /** Helper function for NEON instructions with the format `shrn vd, vn, #imm`.
+   * Ta represents the type of source operand (e.g. for vn.2d, Ta = uint64_t).
+   * Tb represents the type of destination operand (e.g. for vn.2s, Tb =
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/arch/aarch64/helpers/neon.hh]: Updated

**Review by FinnWilkinson — DISMISSED** [review]

File: src/include/simeng/Instruction.hh [hunk 3]
```diff
@@ -33,14 +33,14 @@ class Instruction {
 
   /** Override the specified source register with a renamed physical register.
    */
-  virtual void renameSource(uint8_t i, Register renamed) = 0;
+  virtual void renameSource(uint16_t i, Register renamed) = 0;
 
   /** Override the specified destination register with a renamed physical
    * register. */
-  virtual void renameDestination(uint8_t i, Register renamed) = 0;
+  virtual void renameDestination(uint16_t i, Register renamed) = 0;
```
Comment (dANW34V3R) [inline src/include/simeng/Instruction.hh:40]: Unsure of why this change to the type has been made throughout. How many registers could be contained within e.g. the destinationRegisters vector

Comment (jj16791) [reply src/include/simeng/Instruction.hh:40]: @FinnWilkinson Was this for the SME instructions with matrix destination vectors and implicit registers?

File: src/include/simeng/arch/aarch64/Instruction.hh [hunk 4]
```diff
@@ -51,8 +51,39 @@ inline uint8_t getDataSize(cs_arm64_op op) {
 
   // ARM64_REG_V0 -> {end} are vector registers
   if (op.reg >= ARM64_REG_V0) {
-    // Data size for vector registers relies on opcode thus return 0
-    return 0;
+    // Data size for vector registers relies on opcode, get vector access
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/arch/aarch64/Instruction.hh:54]: We use this function in multiple `.cc` files and could be used in more future developments so it won't be able to move it to a `.cc` file

File: src/include/simeng/Instruction.hh (hunk 3 above)
Comment (FinnWilkinson) [reply src/include/simeng/Instruction.hh:40]: Yes this is for SME. As each row of the ZA matrix is treated like a vector register in SimEng, supposing a 2048-bit vector length you could have up to 256 matrix rows. Add on top of this other operands, you could need up to approx 264 "registers" (i.e. SimEng's internal representation, **NOT** architectural registers)

File: src/include/simeng/Instruction.hh [hunk 5]
```diff
@@ -33,14 +33,14 @@ class Instruction {
 
   /** Override the specified source register with a renamed physical register.
    */
-  virtual void renameSource(uint8_t i, Register renamed) = 0;
+  virtual void renameSource(uint16_t i, Register renamed) = 0;
```
[the remark this answers is not in the document]
Comment (dANW34V3R) [reply src/include/simeng/Instruction.hh:36]: Agreed

File: src/include/simeng/Instruction.hh (hunk 3 above)
Comment (dANW34V3R) [reply src/include/simeng/Instruction.hh:40]: Thanks for the clarification

**Review by dANW34V3R — APPROVED** [review]
Looks good

**Review by FinnWilkinson — APPROVED** [review]

**Review by ABenC377 — APPROVED** [review]
Looks good

**Comment by rahahahat** [comment]
Looks good to me
